crates/sui-adapter/src/execution_mode.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_vm_runtime::session::SerializedReturnValues;
+use sui_types::error::ExecutionError;
+
+pub type TransactionIndex = usize;
+
+pub trait ExecutionMode {
+    type ExecutionResult;
+    type ExecutionResults;
+
+    fn allow_arbitrary_function_calls() -> bool;
+
+    fn make_result(
+        return_values: &SerializedReturnValues,
+    ) -> Result<Self::ExecutionResult, ExecutionError>;
+
+    fn empty_results() -> Self::ExecutionResults;
+
+    fn add_result(
+        results: &mut Self::ExecutionResults,
+        idx: TransactionIndex,
+        result: Self::ExecutionResult,
+    );
+}
+
+pub struct Normal;
+
+impl ExecutionMode for Normal {
+    type ExecutionResult = ();
+    type ExecutionResults = ();
+
+    fn allow_arbitrary_function_calls() -> bool {
+        false
+    }
+
+    fn make_result(srv: &SerializedReturnValues) -> Result<Self::ExecutionResult, ExecutionError> {
+        assert_invariant!(srv.return_values.is_empty(), "Return values must be empty");
+        Ok(())
+    }
+
+    fn empty_results() -> Self::ExecutionResults {
+        ()
+    }
+
+    fn add_result(_: &mut Self::ExecutionResults, _: TransactionIndex, _: Self::ExecutionResult) {
+        ()
+    }
+}
+
+pub struct DevInspect;
```

Comment:
Scary comments warning that this should never be used on a validator would be good